vault/auth_external_plugin_test.go

```diff
@@ -0,0 +1,318 @@
+			panic(err)
+		}
+	})
+
+	// write the cached plugin if necessary
+	var err error
+	if _, err := os.Stat(pluginPath); os.IsNotExist(err) {
+		err = os.WriteFile(pluginPath, pluginBytes, 0o777)
+	}
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	sha := sha256.New()
+	_, err = sha.Write(pluginBytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return pluginName, fmt.Sprintf("%x", sha.Sum(nil)), pluginDir
+}
+
+func TestCore_EnableExternalCredentialPlugin(t *testing.T) {
+	c, pluginName, pluginSha256 := testCoreWithPlugin(t)
+	d := &framework.FieldData{
+		Raw: map[string]interface{}{
+			"name":    pluginName,
+			"sha256":  pluginSha256,
+			"version": "v1.0.0",
+			"command": pluginName,
+		},
+		Schema: c.systemBackend.pluginsCatalogCRUDPath().Fields,
+	}
+	resp, err := c.systemBackend.handlePluginCatalogUpdate(context.Background(), nil, d)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.Error() != nil {
+		t.Fatalf("%#v", resp)
+	}
+	me := &MountEntry{
+		Table:   credentialTableType,
+		Path:    "foo",
+		Type:    pluginName,
+		Version: "v1.0.0",
+	}
+	err = c.enableCredential(namespace.RootContext(nil), me)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	match := c.router.MatchingMount(namespace.RootContext(nil), "auth/foo/bar")
+	if match != "auth/foo/" {
+		t.Fatalf("missing mount, match: %q", match)
+	}
+}
+
+func TestCore_EnableExternalCredentialPlugin_MultipleVersions(t *testing.T) {
+	c, pluginName, pluginSha256 := testCoreWithPlugin(t)
+	d := &framework.FieldData{
+		Raw: map[string]interface{}{
+			"name":    pluginName,
+			"sha256":  pluginSha256,
+			"version": "v1.0.0",
+			"command": pluginName,
+		},
+		Schema: c.systemBackend.pluginsCatalogCRUDPath().Fields,
+	}
+	resp, err := c.systemBackend.handlePluginCatalogUpdate(context.Background(), nil, d)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.Error() != nil {
+		t.Fatalf("%#v", resp)
+	}
+
+	d = &framework.FieldData{
+		Raw: map[string]interface{}{
+			"name":    pluginName,
+			"sha256":  pluginSha256,
+			"version": "v1.0.1",
+			"command": pluginName,
+		},
+		Schema: c.systemBackend.pluginsCatalogCRUDPath().Fields,
+	}
+	resp, err = c.systemBackend.handlePluginCatalogUpdate(context.Background(), nil, d)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if resp.Error() != nil {
+		t.Fatalf("%#v", resp)
+	}
+
+	me := &MountEntry{
+		Table:   credentialTableType,
+		Path:    "foo",
+		Type:    pluginName,
+		Version: "v1.0.0",
+	}
+	err = c.enableCredential(namespace.RootContext(nil), me)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+
+	match := c.router.MatchingMount(namespace.RootContext(nil), "auth/foo/bar")
+	if match != "auth/foo/" {
+		t.Fatalf("missing mount, match: %q", match)
+	}
+
+	raw, _ := c.router.root.Get(match)
+	if raw.(*routeEntry).mountEntry.Version != "v1.0.0" {
+		t.Errorf("Expected mount to be version v1.0.0 but got %s", raw.(*routeEntry).mountEntry.Version)
+	}
+}
+
+func TestCore_EnableExternalCredentialPlugin_MismatchedVersions(t *testing.T) {
```

Comment:
The behaviour in this test looks correct to me, but the name seems a bit odd, does MultipleVersions make more sense than MismatchedVersions, or am I misreading? Is it essentially the same as the previous test, but selecting 1.0.1 for mounting instead of 1.0.0?

Reply:
Multiple versions is definitely better. And yes, you are correct as far as the behavior.